### PR TITLE
Bluetooth: improved pairing process and host layer now successfully receives ACL packets

### PR DESCRIPTION
**Pull request by robertc2000**
## Summary

This PR aims to improve and fix the core functionalities of the Bluetooth stack.

Initially, I tried to connect to the NuttX device (a ESP32 device using the esp32-devkitC:ble menu configuration) via Bluetooth and attempted to read the services/characteristics that it was contains. I noticed that the connection did not work, because the NuttX device (the host layer) was not receiving  any ACL packets (containing my requests for these services) from the controller layer. Therefore, the device running NuttX was not replying to my requests and connection failed.

I found a fix for this issue: the 'Controller to Host Flow Control' parameter has to be set to 0. This fixed the issue of packets not arriving in the host layer. With this setup, I got a successful connection. I decided to create an option in the Bluetooth menu, in which the user can select the value for this 'Controller to Host Flow Control' parameter. The reason for this is because I only tested on ESP32 and I got successful connection with the value set to 0. I didn't test on other devices, therefore I cannot confirm this value will work for any non-ESP32, so it's best to keep this parameter user-configurable.

Another improvement I made was regarding the pairing process. I implemented the handling of the LL_CONNECTION_PARAM_REQ event and I noticed a considerable response speed improvement from the NuttX stack. Previously, the stack was using the 'bt_l2cap_update_conn_param' for connection parameter update, however I noticed the stack was a bit laggy in responding, so I commented this out for the moment.

## Testing

I used a ESP32-DEVKITC board along with the esp32-devkitc:ble configuration. To connect to the device, I used my smartphone along with an application for connecting using Bluetooth ('nRF connect').

I also discovered some bugs in the process:
1. Advertising does not work, unless scanning is enabled. I used the 'btsak' app from the NuttX terminal (nsh) in order to enable advertising and scanning, and I noticed that enabling advertising does nothing. Only after enabling scanning, the device starts to also advertise.

2. Transmission of packets does not work (from the host layer: L2CAP, GATT, etc..), unless logging is done simultaneously. I suspect there might be some synchronization/race condition error in the scheduler, because a separate thread is used for Bluetooth TX (it might not be necessarily related to the BLE stack).

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)

__Fill In The Commit Message:__ This PR contains a Commit with an Empty Commit Message. Please fill in the Commit Message with the PR Summary.



## NuttX PR Requirements Review

This PR summary provides a good starting point, but it needs some work to meet all the NuttX requirements. Here's a breakdown:

**Strengths:**

* **Clear explanation of the problem:** The summary clearly identifies the issue with ACL packets not reaching the host layer and the connection issues it causes.
* **Solution description:**  The fix of setting the 'Controller to Host Flow Control' parameter is well-explained, including the reasoning behind making it user-configurable. 
* **Additional improvement:** The improvement to the pairing process by handling `LL_CONNECTION_PARAM_REQ` is a valuable addition.
* **Testing details:** You mention the target board and configuration used for testing, along with the tool used for connection.

**Areas for Improvement:**

* **Summary:**
    * **Why is the change necessary?**  While the problem is stated, explicitly mention the impact on users (e.g., inability to connect to Bluetooth devices reliably).
    * **What functional part of the code is being changed?**  Be more specific than "Bluetooth stack."  Mention specific files or modules impacted (e.g., HCI layer, L2CAP).
    * **Issue references:** Include links to any relevant NuttX or NuttX Apps issues or pull requests.
* **Impact:**
    * **New feature:** The configurable 'Controller to Host Flow Control' is a new feature. State this explicitly. 
    * **User impact:** Users with non-ESP32 devices may need to experiment with this parameter to achieve successful connections.
    * **Hardware impact:**  Specify that the changes primarily affect Bluetooth-enabled boards. 
    * **Documentation:**  Documentation will need updating to describe the new configuration option and its implications. 
* **Testing:**
    * **Build Host(s):**  Provide details about your development environment (OS, CPU, compiler).
    * **Testing logs:** The provided logs are placeholders. Include actual logs showing the behavior before and after the changes.
    * **Bug reports:**
        * The issues with advertising and transmission seem significant. 
        * Each bug should be reported as a separate issue on the NuttX issue tracker with detailed descriptions and steps to reproduce. 
        * Consider addressing these bugs in separate PRs for better clarity and review. 

**Conciseness:**

* The summary is a bit long. Try to be more concise while still providing all necessary information. 

**Overall:**

The PR addresses important issues and introduces useful improvements to the Bluetooth stack. However, you need to provide more specific details, complete the required sections, and address the identified bugs to ensure a smooth review and merge process.

**Comment by acassis** [comment]
Hi @robertc2000 this fix is very important! NuttX needs better support to BLE!!!

@phfbertoleti @tmedicci @eren-terzioglu @fdcavalcanti please take a look

**Comment by acassis** [comment]
@robertc2000 please fix the coding style issues:

```
Error: /home/runner/work/nuttx/nuttx/nuttx/wireless/bluetooth/bt_hcicore.c:746:6: error: C++ style comment
Error: /home/runner/work/nuttx/nuttx/nuttx/wireless/bluetooth/bt_hcicore.c:910:54: error: Left bracket not on separate line
Error: /home/runner/work/nuttx/nuttx/nuttx/wireless/bluetooth/bt_hcicore.c:915:93: error: Long line found
Error: /home/runner/work/nuttx/nuttx/nuttx/wireless/bluetooth/bt_hcicore.c:930:43: error: Upper case hex constant found
Error: /home/runner/work/nuttx/nuttx/nuttx/wireless/bluetooth/bt_hcicore.c:932:85: error: Long line found
Error: /home/runner/work/nuttx/nuttx/nuttx/wireless/bluetooth/bt_hcicore.c:2131:57: error: Dangling whitespace at the end of line
```

You can use the nxstyle tool or the checkpatch.sh (that uses the nxstyle) to find issue before creating the commit, more info: https://nuttx.apache.org/docs/latest/contributing/index.html

File: wireless/bluetooth/Kconfig [hunk 1]
```diff
@@ -165,6 +165,16 @@ config BLUETOOTH_BUFFER_IRQRESERVE
 		interrupt level.  This setting only needs to be non-zero if your
 		low-level Bluetooth driver needs to do such allocations.
 
+config BLUETOOTH_CNTRL_HOST_FLOW_ENABLE
+	bool "Enable Controller to Host Flow Control"
+	default n if ARCH_CHIP_ESP32
```
Comment (acassis) [inline wireless/bluetooth/Kconfig:170]: @robertc2000 actually we don't include arch specific rules here. Case you need to control it this way, it is possible to create an intermediate symbol (i.e. BLE_HOST_HAS_FLOW_CONTROL) and in the arch/esp32 you enable the BLE_HOST_HAS_FLOW_CONTROL

File: arch/xtensa/src/esp32/Kconfig [hunk 2]
```diff
@@ -2194,6 +2194,10 @@ config ESP32_BLE_MAX_CONN
 	range 1 3
 	default 1
 
+config BT_CTRL_HOST_FLOW_CONTROL_SUPPORT_DISABLED
+    bool
+    default y
+
```
Comment (acassis) [inline arch/xtensa/src/esp32/Kconfig:2200]: Please include ---help--- message. Explain clearly what is this feature and why is this needed?

Comment (acassis) [reply arch/xtensa/src/esp32/Kconfig:2200]: Oops, sorry, it is an intermediate flag that is not visible to the user, please ignore my previous comment

File: wireless/bluetooth/Kconfig [hunk 3]
```diff
@@ -165,6 +165,16 @@ config BLUETOOTH_BUFFER_IRQRESERVE
 		interrupt level.  This setting only needs to be non-zero if your
 		low-level Bluetooth driver needs to do such allocations.
 
+config BLUETOOTH_CNTRL_HOST_FLOW_ENABLE
+	bool "Enable Controller to Host Flow Control"
+	default n if BT_CTRL_HOST_FLOW_CONTROL_SUPPORT_DISABLED
+	default y
```
Comment (acassis) [inline wireless/bluetooth/Kconfig:171]: I think you can replace both with just:
```
default !BT_CTRL_HOST_FLOW_CONTROL_SUPPORT_DISABLED
```

This way it will be disabled when BT_CTRL_HOST_FLOW_CONTROL_SUPPORT_DISABLED is enabled

File: wireless/bluetooth/Kconfig [hunk 4]
```diff
@@ -165,6 +165,16 @@ config BLUETOOTH_BUFFER_IRQRESERVE
 		interrupt level.  This setting only needs to be non-zero if your
 		low-level Bluetooth driver needs to do such allocations.
 
+config BLUETOOTH_CNTRL_HOST_FLOW_ENABLE
+	bool "Enable Controller to Host Flow Control"
+	default n if BT_CTRL_HOST_FLOW_CONTROL_SUPPORT_DISABLED
```
Comment (raiden00pl) [inline wireless/bluetooth/Kconfig:170]: this way you still rely on the arch specific flag.
You should `select BLUETOOTH_CNTRL_HOST_FLOW_ENABLE` in your arch specific `Kconfig` when BLE is enabled.

Comment (raiden00pl) [reply wireless/bluetooth/Kconfig:170]:
````
config ESP32_BLE
	bool "BLE"
	default n
	select ESP32_WIRELESS
	select BLUETOOTH_TXCMD_PINNED_TO_CORE if SMP
        select BLUETOOTH_CNTRL_HOST_FLOW_ENABLE if WIRELESS_BLUETOOTH
```

Comment (acassis) [reply wireless/bluetooth/Kconfig:170]: Good point @raiden00pl !!! @robertc2000 actually that flag BT_CTRL_HOST_FLOW_CONTROL_SUPPORT_DISABLED should be defined wireless/bluetooth/Kconfig as disabled by default and at arch (like ESP32 Kconfig) that flag should be enabled

Comment (robertc2000) [reply wireless/bluetooth/Kconfig]: Actually, for ESP32, the flow should be disabled, so BLUETOOTH_CNTRL_HOST_FLOW_ENABLE should be off only for it.

I agree though, that's a good point! Thanks for the hint! Pushed the changes now which correct that.

**Review by acassis — APPROVED** [review]
